### ISSUE TITLE
RustTypeInferencingVisitor is getting out of hand

### ISSUE THREAD

**Issue by matklad** [intellij-rust/intellij-rust#487]
Our [RustTypeInferencingVisitor] is a bit hard to follow. There is a number of casts and a bit of duplicated code in each of the visiting functions. I think that one of the problems is that the `path` is very weakly typed: it's just a `LinkedList<PsiElement>`. @alexeykudinkin  what do you think about introducing an abstraction for the path to the particular binding inside a pattern? I've tried to do some experiments [here].

[RustTypeInferencingVisitor]: https://github.com/intellij-rust/intellij-rust/blob/master/src/main/kotlin/org/rust/lang/core/types/visitors/RustTypeInferenceEngine.kt#L24

[here]: https://github.com/intellij-rust/intellij-rust/commit/f6a7102a73cf0b7cab16ba2eb1fd131add4ba7ad#diff-04a810d0411b4b04c937b207f054a71cR23

**Comment by alexeykudinkin** [comment]
Agree, it's ugly in its current form, though i don't believe that we need yet another abstraction. What we need actually is just to organise it in the right way (as a proper automaton for ex).

**Comment by matklad** [comment]
Please note that `pat_struct` can be applied to both structs and enum variants, if they are struct like. This is an important bit that is missed from the current implementation.

**Comment by matklad** [comment]
I am trying to say that [this](https://github.com/intellij-rust/intellij-rust/blob/master/src/test/kotlin/org/rust/lang/core/type/RustEnumPatternTypeInferenceTest.kt#L40) should be `<unknown>`, and not `i64`.

**Comment by alexeykudinkin** [comment]
As, far as i see, grammar doesn't allow named pats in `pat_enum`.

**Comment by matklad** [comment]
But it allows them in `pat_struct`, so the test should be written as

```Rust
        enum E {
            X { _1: i32, _2: i64 }
        }
        fn bar() -> E {}
        fn main() {
            let E::X { _2: i, ..} = bar();
            i;
          //^ i64
```

(and perhaps we want to tweak the grammar as well)
